src/libcore/mem/mod.rs

```diff
@@ -88,16 +109,20 @@ pub use crate::intrinsics::transmute;
 /// // `s` is implicitly dropped and its memory deallocated.
 /// ```
 ///
-/// Using `ManuallyDrop` here has two advantages:
+/// `ManuallyDrop` robustly prevents double-free because we disable `v`'s destructor
+/// before doing anything else. `mem::forget()` doesn't allow this because it consumes its
+/// argument, forcing us to call it only after extracting anything we need from `v`.
+///
+/// If a panic were introduced between construction of `ManuallyDrop` and building the
+/// string (which cannot happen in the code as shown), it would result in a leak and not a
+/// double free. In other words, `ManuallyDrop` errs on the side of leaking instead of
+/// erring on the side of dropping.
 ///
-/// * We do not "touch" `v` after disassembling it. For some types, operations
-///   such as passing ownership (to a funcion like `mem::forget`) requires them to actually
-///   be fully owned right now; that is a promise we do not want to make here as we are
-///   in the process of transferring ownership to the new `String` we are building.
-/// * In case of an unexpected panic, `ManuallyDrop` is not dropped, but if the panic
-///   occurs before `mem::forget` was called we might end up dropping invalid data,
-///   or double-dropping. In other words, `ManuallyDrop` errs on the side of leaking
-///   instead of erring on the side of dropping.
+/// Finally, `ManuallyDrop` prevents us from having to "touch" `v` after transferring the
+/// ownership to `s`. For some types, referencing dangling data or no longer owning it
+/// means the value is invalid and that passing it to a function (even one like
+/// `mem::forget`) constitutes undefined behavior. With `ManuallyDrop` final step to get
+/// rid of the now invalid object is avoided.
```

Comment:
That last sentence seems to be ungrammatical, I cannot entirely figure out what it is that you wanted to say. Maybe "*the* final step"?